rfc/rfc000-package-server.md

```diff
@@ -1,19 +1,91 @@
+# Package Server
+
 ---
+
 RFC: 000
-Author: Peter Bourgon <peter@bourgon.org>
+
+Author: Matt Heath <matt@mattheath.com>
+
 Status: Draft
----
 
-# package server
+---
 
 ## Motivation
 
-http://peter.bourgon.org/go-kit/#package-server
+[http://peter.bourgon.org/go-kit/#package-server](http://peter.bourgon.org/go-kit/#package-server)
+
+Package server is probably the biggest and most important component of the toolkit. Ideally, we should be able to write our services as implementations of normal, nominal Go interfaces, and delegate integration with the environment to the server package. The package should encode and enforce conventions for server-side concerns, like health checks, system-wide request tracing, connection management, backpressure and throttling, and so on. For each of those topics, it should provide interfaces for different, pluggable strategies. It should integrate with service discovery, and work equally well over multiple transports. Considerable prior art exists in the form of Finagle, Karyon (Netflixs application service library), and likely many more.
 
 ## Scope
 
-To be defined.
+### Endpoints
+
+*   An endpoint is defined as a handler interface which receives a request and returns a response or an error to the client.
```

Comment:
Unrelated to the context stuff: I often wish that we could support schemes other than a single Request/Response pair. I.e., it would be great if we had built-in support from day one for channels; certainly for the responses, but perhaps also for the requests, thus allowing for generic message-passing between clients and servers.

To motivate this, consider a SaaS system that has an embedded client library. Due to the way that various VPCs/NATs are configured, it is of course not generally possible for the SaaS system to open a connection to its clients directly. So, instead, the clients do a "long-poll" pattern to gather information from the server. It would be cleaner in my mind if the client could initiate the connection (required due to the aforementioned networking issues) but, after the connection is initiated, it's just a bidirectional channel between client and server.

(If anyone is swayed by this, we could discuss exactly how the interface would look, how Context would be treated, etc)